### PR TITLE
Add SumConstraint projection MPO

### LINKED ISSUE #58: Implement SumConstraint projection MPO
Parent epic: #55

Planned PR branch/name: `pr-3-sum-constraint-projection`

## Scope
- Implement CoTenN-style automaton lowering for weighted binary sums.
- Support senses `:<=`, `:<`, `:==`, `:>=`, and `:>`.
- Use nonnegative integer weights and integer RHS.
- Add exact projection-mask tests for small systems.

## Acceptance Criteria
- [ ] `projection_mpo(SumConstraint(...), sites)` matches brute-force feasibility masks.
- [ ] Supports non-contiguous sites using identity/pass-through behavior.
- [ ] Tests cover the knapsack-style example from the CoTenN artifact.
- [ ] Invalid unsupported domains fail clearly.

## Depends On
- #57 (Add native projection-MPO assembly utilities)

## Notes
Each issue is intended to be completed by one focused PR.
## Attribution
This work is inspired by the projection-MPO constraint framework in Sharma, Ritvik, Cheng Peng, Siddharth Dangwal, and Sara Achour. "CoTenN: Constrained Optimization with Tensor Networks." Proceedings of the ACM on Programming Languages 10, no. PLDI (2026): 629-655.

## Benchmark Follow-Up
- The knapsack penalty-QUBO vs projection-MPO comparison is tracked in #66. This SumConstraint work provides the hard-constraint projection used by that benchmark.

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

- Adds SumConstraint-specific projection MPO lowering using reachable integer partial-sum states.
- Preserves pass-through behavior for unconstrained sites and supports the accepted v1 relations: `==`, `!=`, `<=`, and `>=`.
- Validates integer-valued SumConstraint projection weights/RHS and documents exact `==`/`!=` comparison behavior.
- Adds exact projection-mask tests for all accepted SumConstraint relations, non-contiguous sites, unsupported domains, overflow guards, pruning, and a small knapsack-style capacity mask.
- Incorporates the projection-MPO core work from #85 on `main`, including the shared `AbstractConstraint` dispatch, documentation depth, `SparseTensorEntry` semantics, and coefficient anchoring.

Closes #58.

## Tests

- `git diff --check` passed.
- `julia +release --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 273/273.
- `julia +release --project=. -e 'using Pkg; Pkg.test()'` passed.
- GitHub Actions on `6f24146` passed:
  - Documentation `build`
  - Run tests matrix: Julia lts/1/pre on Ubuntu, Windows, and macOS
  - `documenter/deploy`

## Branch Hygiene

- Default branch: `main`.
- Base branch: `main`.
- Current base incorporated: `origin/main` at `15f52d383f97bd8c10272a749059217da52dfe94`.
- Latest merge-resolution commit: `6f24146` merges `origin/main` into `pr-3-sum-constraint-projection` and resolves the projection MPO/test conflicts after #85 landed.
- Stacked status: no longer stacked; prerequisite #85 has been incorporated on `main`.
- Current PR status: ready for review, merge-clean, with all checks passing on the latest head.

### CODE REVIEW

File: src/projection_mpo.jl [hunk 1]
```diff
@@ -188,14 +201,162 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint, constrain
...
 
 Build a diagonal projection MPO that preserves basis states satisfying
 `constraint` and zeros infeasible states.
+
+`SumConstraint` projection MPOs use exact integer partial sums, so their
+weights and right-hand side must be integer-valued.
 """
 function projection_mpo(::Type{T}, constraint::SumConstraint, sites) where {T}
-  return projection_mpo(T, constraint, constraint_sites(constraint), sites)
+  return sum_constraint_projection_mpo(T, constraint, collect(sites))
```
Comment (bernalde) [inline src/projection_mpo.jl]: Nonblocking (dead code): rerouting `SumConstraint` through `sum_constraint_projection_mpo` makes the `SumConstraint` overloads of `projection_entries` (line 167) and `constraint_sites` (line 151) unreachable. They were the live path for `SumConstraint` before this PR; the other constraint types still use them, so only these two `SumConstraint`-typed methods are now dead. Remove them, or add a comment noting they are intentionally retained. An unused parallel path is the kind of drift @iagoleal flagged on #84.

File: src/projection_mpo.jl [hunk 2]
```diff
@@ -188,14 +201,162 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint, constrain
...
+end
+
+function sum_projection_reachable_states(weights_by_site, num_sites)
+  states_after_site = Vector{Vector{Int}}(undef, num_sites)
+  reachable = [0]
+
+  for site_position in 1:num_sites
+    weight = get(weights_by_site, site_position, 0)
+    shifted = [checked_sum_projection_add(partial_sum, weight) for partial_sum in reachable]
+    next_reachable = unique!(sort!(vcat(reachable, shifted)))
```
Comment (bernalde) [inline src/projection_mpo.jl]: Nonblocking + question (cost / bond dimension): the interior bond dimension equals the number of distinct reachable partial sums, which is exponential for weights like powers of two. I verified `maxlinkdim == 32` for `weights = [1,2,4,8,16,32]` at n=6, versus 6 for unit weights at capacity 3. This is inherent to exact partial-sum projection (and matches CoTenN), but the `projection_mpo` docstring only says weights must be integer-valued; it does not warn that memory/cost scales with the number of distinct partial sums. Please add a sentence to the docstring.

Separately: for `<=`/`==` with nonnegative weights, partial sums exceeding `rhs` are dead yet still tracked here. Pruning them would bound the bond dimension by `rhs + 1` for those relations. Intended to keep the code uniform, or worth pruning?

File: test/projection_mpo.jl [hunk 3]
```diff
@@ -43,19 +55,52 @@ end
     @test_throws ArgumentError TenSolver.tensor_to_mpo(Float64, [], ITensors.Index{Int64}[])
     @test_throws ArgumentError TenSolver.tensor_to_mpo(Float64, [], qutrit_sites)
     @test_throws BoundsError TenSolver.projection_mpo(ExactlyOneConstraint([4]), sites)
+    @test_throws ArgumentError TenSolver.projection_mpo(
+      SumConstraint([1], [1], Symbol("<="), 1),
+      qutrit_sites,
+    )
+    @test_throws ArgumentError TenSolver.projection_mpo(
+      SumConstraint([1], [0.5], Symbol("<="), 1),
```
Comment (bernalde) [inline test/projection_mpo.jl]: Nonblocking (test gap): the integer-validation errors are covered well, but the overflow guards this PR adds are not exercised: `integer_projection_value`'s Int-conversion `InexactError`/`OverflowError` branch, and `checked_sum_projection_add`'s `OverflowError` branch. They are defensive and unlikely in practice. Either add a case (e.g. a weight beyond `typemax(Int)` via `BigInt`, or weights summing past `typemax(Int)`) or leave them as deliberate defensive paths and note so. Not a gate.

**Review by bernalde — COMMENTED** [review]
Senior maintainer review of PR #86 (SumConstraint projection MPO).

Verdict mechanics: I am the PR author (authenticated as `bernalde`), so GitHub rejects a formal APPROVE/REQUEST_CHANGES from this account (HTTP 422). Submitting as COMMENT. Substantive verdict is approve-equivalent: no blocking issues. Formal approval must come from another maintainer (e.g. @iagoleal).

Review setup:
- Fetched `refs/pull/86/head` (`2dc51ae`) and base branch `pr-2-projection-mpo-core` (`b65bb83`, = PR #85 head). This PR is stacked on the still-open #85; merge-base is `b65bb83`.
- Reviewed a clean detached worktree at the PR head.
- Local diff vs merge-base matches GitHub's reported totals: 3 files, +221 / -12.

Intent vs issue #58:
Delivers the issue scope. `projection_mpo(SumConstraint(...), sites)` builds a partial-sum automaton MPO, and the tests assert it matches `is_feasible` brute force for `==`, `!=`, `<=`, `>=`; non-contiguous sites pass through via zero weights; a knapsack-style capacity mask is covered; unsupported domains (qutrit) and non-integer weights/rhs raise `ArgumentError`. `Closes #58` is appropriate. Note: merging into the non-default base `pr-2-projection-mpo-core` will not auto-close #58 until the stack reaches `main`.

Correctness: I traced the automaton. Link position p at each cut maps consistently to `states_after_site[i]` between adjacent tensors; every transition target is in the next reachable set (no missing key); tensors are physically diagonal; the relation filter is applied only at the final site, on the total sum. Verified empirically that diagonal entries are exactly 0/1 and match feasibility for all tested relations.

Blocking: none.

Nonblocking:
1. Dead code: rerouting SumConstraint to the automaton leaves the SumConstraint overloads of `projection_entries` and `constraint_sites` unreachable. See inline.
2. Bond-dimension growth with no docstring warning (exponential for power-of-two weights). See inline.
3. Overflow guards added by this PR are untested. See inline.

Questions:
1. Issue #58's scope text lists strict `<`/`>` senses, which `SumConstraint` intentionally does not support (removed in #84). The PR correctly implements only `==`/`!=`/`<=`/`>=`. Confirming this is a deliberate match to the merged relation set and the #58 text is simply stale (consistent with my earlier note on #58); `Closes #58` remains correct.
2. For `<=`/`==` with nonnegative weights, partial sums exceeding `rhs` can never become feasible but are still tracked as reachable states. Intentional for code uniformity, or worth pruning to bound the bond dimension? See inline 2.

Tests run (clean worktree, Julia 1.10.11):
- `julia --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` -> Projection MPOs 129/129 pass.
- `julia --project=. -e 'using Pkg; Pkg.test()'` -> full suite passed ("Testing TenSolver tests passed"): Projection MPOs 129, QUBODrivers 169, Aqua 11, Doctests 1, HDF5 16. The 3 "Broken" Real Models entries are pre-existing `@test_broken`, unrelated to this PR.

Merge-readiness: Mergeable on the merits; no blocking issues. Must not merge ahead of its base PR #85, and should be retargeted to `main` once #85 lands, as the PR body states. The nonblocking items are cleanups, not gates. Formal approval must come from another maintainer since I authored the PR.

**Comment by bernalde** [comment]
Commits pushed:
- `184f9fa` `fix: address SumConstraint projection review notes`

Main changes made:
- Removed the now-dead `SumConstraint` overloads for `constraint_sites` and `projection_entries`.
- Added a `projection_mpo` docstring warning that SumConstraint link dimensions grow with the number of distinct reachable partial sums and can be exponential.
- Added overflow coverage for too-large integer projection values and partial-sum overflow during automaton construction.

Tests run:
- `julia --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 131/131.
- `julia --project=. -e 'using Pkg; Pkg.test()'` passed.
- `git diff --check` passed.
- No separate lint/type-check command was found beyond the Julia test/Aqua coverage in `Pkg.test()`.

Comments intentionally not addressed:
- Relation-specific pruning for `<=`/`==` was not implemented in this PR. I kept the current uniform exact automaton and documented its cost; pruning is a performance optimization that should be handled as a follow-up if desired.

Remaining risks or follow-up items:
- PR #86 remains stacked on #85 and should be retargeted after #85 lands.
- Optional follow-up: relation-specific partial-sum pruning to reduce bond dimensions for nonnegative `<=`/`==` constraints.

File: src/projection_mpo.jl (hunk 1 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in https://github.com/SECQUOIA/TenSolver.jl/commit/184f9fa192a57fae451be7a7c6789810423142e6 by removing the dead `SumConstraint` `constraint_sites` and `projection_entries` overloads. The non-sum constraints still use the generic entry path.

File: src/projection_mpo.jl (hunk 2 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Partially addressed in https://github.com/SECQUOIA/TenSolver.jl/commit/184f9fa192a57fae451be7a7c6789810423142e6: added the `projection_mpo` docstring warning about link-dimension growth with distinct partial sums. I did not implement relation-specific pruning here; I kept this PR to the uniform exact automaton and left pruning as a follow-up performance change.

File: test/projection_mpo.jl (hunk 3 above)
Comment (bernalde) [reply test/projection_mpo.jl]: Addressed in https://github.com/SECQUOIA/TenSolver.jl/commit/184f9fa192a57fae451be7a7c6789810423142e6 by adding tests for a `BigInt` weight that cannot fit in `Int` and for partial-sum overflow past `typemax(Int)`. `test/projection_mpo.jl`; projection tests now pass 131/131.

File: src/projection_mpo.jl (hunk 2 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Leaving this thread open to track what is still outstanding. The documentation ask is done (docstring now warns that link dimensions grow with the number of distinct reachable partial sums). Not yet done: the relation-specific pruning. For `<=`/`==` with nonnegative weights, partial sums exceeding `rhs` are still tracked as reachable states even though they can never become feasible; pruning them would bound the interior bond dimension by `rhs + 1` for those relations. This was explicitly deferred as a follow-up performance change and is not a gate for this PR, so this thread stays open only to track that follow-up.

**Comment by bernalde** [comment]
Commits pushed:
- `b754dd8` `perf: prune SumConstraint projection states`

Main changes made:
- Implemented relation-specific partial-sum pruning for nonnegative `<=` and `==` SumConstraint projection MPOs.
- Dropped transitions whose next partial sum would exceed `rhs` before they become link states.
- Kept unbounded exact automaton behavior for `>=` and `!=`.
- Updated the projection docstring to state that `<=` and `==` prune partial sums above the right-hand side.
- Added tests for the power-of-two example, checking both feasibility masks and `maxlinkdim <= rhs + 1` for capacity 3.

Tests run:
- `julia --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 261/261.
- `julia --project=. -e 'using Pkg; Pkg.test()'` passed.
- `git diff --check` passed.

Comments intentionally not addressed:
- None.

Remaining risks or follow-up items:
- PR #86 remains stacked on #85 and should be retargeted after #85 lands.

File: src/projection_mpo.jl (hunk 2 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in https://github.com/SECQUOIA/TenSolver.jl/commit/b754dd829a9d61eb8ddab4bcce2cfbb416f9c4c6. `<=` and `==` SumConstraint projection now prune transitions whose partial sum would exceed `rhs`, so the power-of-two capacity-3 case stays within `rhs + 1` link states. Added feasibility-mask and `maxlinkdim <= 4` tests in `test/projection_mpo.jl`; projection tests pass 261/261.

**Comment by bernalde** [comment]
Updated this stacked PR with the latest #85 base changes.

Commits pushed:
- `605b695` `Merge projection MPO core updates into SumConstraint projection`

What changed:
- Merged `origin/pr-2-projection-mpo-core` at `4137ccd`.
- Resolved the projection conflict by keeping the generic `projection_mpo(::AbstractConstraint, sites)` path for non-sum constraints and the narrow `SumConstraint` partial-sum automaton specialization for #58.
- Preserved #85's review follow-ups: shared `constraint_sites`, deeper projection docs, `SparseTensorEntry` documentation, and coefficient anchoring on the first constrained site.
- Updated the PR body Branch Hygiene section and local verification list.

Verification:
- `julia +release --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 273/273.
- `julia +release --project=docs -e 'using Pkg; Pkg.instantiate(); include("docs/make.jl")'` passed; Documenter skipped deployment because this was a local build.

Remaining:
- GitHub Actions are running for the new head.
- This PR remains draft and stacked on #85.

**Comment by bernalde** [comment]
Commits pushed:
- `6f24146` `Merge main into SumConstraint projection PR`

Main changes made:
- Merged current `origin/main` (`15f52d3`) into `pr-3-sum-constraint-projection`.
- Resolved the projection MPO and projection test conflicts by keeping `main`'s generic helper refactor and preserving the SumConstraint automaton, pruning, validation, and tests from this PR.
- Marked the PR ready for review now that it is no longer stacked on #85.
- Updated the PR body to reflect `main` as the base branch, the latest verification, and the current merge-clean status.

Tests run:
- `git diff --check` passed.
- `julia +release --project=. -e 'using Test, TenSolver; include("test/projection_mpo.jl")'` passed: Projection MPOs 273/273.
- `julia +release --project=. -e 'using Pkg; Pkg.test()'` passed.
- GitHub Actions on `6f24146` passed: Documentation, full Run tests matrix, and `documenter/deploy`.

Comments intentionally not addressed:
- None. There are no unresolved inline review threads.

Remaining risks or approval gates:
- PR #86 is ready for review, merge-clean, and green, but still needs formal maintainer approval before merge.

File: src/projection_mpo.jl [hunk 4]
```diff
@@ -124,15 +124,17 @@ function projection_site_tensor(::Type{T}, site, left_link, right_link, entries,
   end
 end
 
+function validate_projection_sites(sites)
+  isempty(sites) && throw(ArgumentError("sites must not be empty"))
+  all(site -> ITensors.dim(site) == 2, sites) ||
+    throw(ArgumentError("projection MPO sites must be binary Qudit indices"))
```
Comment (iagoleal) [inline src/projection_mpo.jl]:
```suggestion
  if isempty(sites)
    throw(ArgumentError("sites must not be empty"))
  end
  if !all(site -> ITensors.dim(site) == 2, sites)
    throw(ArgumentError("projection MPO sites must be binary Qudit indices"))
  end

File: src/projection_mpo.jl [hunk 5]
```diff
@@ -174,6 +175,13 @@ function validate_sparse_entries(entries, sites)
   return nothing
 end
 
+function validate_constraint_site_bounds(constraint_sites, sites)
+  all(site -> 1 <= site <= length(sites), constraint_sites) ||
+    throw(BoundsError(sites, maximum(constraint_sites)))
```
Comment (iagoleal) [inline src/projection_mpo.jl]:
```suggestion
  if !all(site -> 1 <= site <= length(sites), constraint_sites)
    throw(BoundsError(sites, maximum(constraint_sites)))
  end
```

File: src/projection_mpo.jl [hunk 6]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+  catch error
+    if error isa OverflowError
+      throw(ArgumentError("SumConstraint projection partial sums must fit in Int"))
+    end
+    rethrow()
+  end
+end
+
+function sum_projection_state_limit(relation::Symbol, rhs::Int)
+  (relation === Symbol("<=") || relation === Symbol("==")) && return rhs
```
Comment (iagoleal) [inline src/projection_mpo.jl]:
```suggestion
  if (relation === Symbol("<=") || relation === Symbol("=="))
    return rhs
  end
```

File: src/projection_mpo.jl [hunk 7]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+
+function sum_transition_target(partial_sum::Int, weight::Int, bit::Integer, state_limit)
+  bit == 0 && return partial_sum
+
+  if !isnothing(state_limit) && sum_exceeds_state_limit(partial_sum, weight, state_limit)
+    return nothing
+  end
+
+  return checked_sum_projection_add(partial_sum, weight)
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: Prefer a if-elseif-else chain in here for readability.

File: src/projection_mpo.jl [hunk 8]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+    states_after_site[site_position] = next_reachable
+    reachable = next_reachable
+  end
+
+  return states_after_site
+end
+
+function sum_state_position_map(states)
+  return Dict(state => position for (position, state) in pairs(states))
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: This is small and only used once. We can inline it on the call site.

File: src/projection_mpo.jl [hunk 9]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+
+  try
+    return Int(value)
+  catch error
+    if error isa InexactError || error isa OverflowError
+      throw(ArgumentError("$name must fit in Int for SumConstraint projection MPOs"))
+    end
+    rethrow()
+  end
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: A better idea than to keep validating if the weights are integer is to use the type system capabilities and only accept constraints of type `SumConstraint{<:Integer}` in all methods.

File: src/projection_mpo.jl [hunk 10]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+
+function mpo_transition_coordinate(state::Integer, left_position, right_position, left_link, right_link)
+  coordinate = Int[]
+  !isnothing(left_link) && push!(coordinate, left_position)
+  push!(coordinate, state, state)
+  !isnothing(right_link) && push!(coordinate, right_position)
+
+  return coordinate
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: This is almost identical to `mpo_coordinate` and can be unified with it in a single method.
That is, substitute all calls to `mpo_coordinate` with calls to `mpo_transition_coordinate` where both positions are the same.

Comment (iagoleal) [inline src/projection_mpo.jl]: This is almost identical to `mpo_coordinate` and can be unified with it in a single method.

File: src/projection_mpo.jl [hunk 11]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+
+function checked_sum_projection_add(lhs::Int, rhs::Int)
+  try
+    return Base.checked_add(lhs, rhs)
+  catch error
+    if error isa OverflowError
+      throw(ArgumentError("SumConstraint projection partial sums must fit in Int"))
+    end
+    rethrow()
+  end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: I appreciate the improved error message here, but it is still an `OverflowError`, not an `ArgumentError`.

File: src/projection_mpo.jl [hunk 12]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+function sum_constraint_projection_data(constraint::SumConstraint)
+  weights_by_site = Dict{Int,Int}()
+  for (site, weight) in zip(constraint.sites, constraint.weights)
+    weights_by_site[site] = integer_projection_value(weight, "SumConstraint weights")
+  end
+
+  rhs = integer_projection_value(constraint.rhs, "SumConstraint rhs")
+
+  return weights_by_site, rhs
+end
```
Comment (iagoleal) [inline src/projection_mpo.jl]: I like this conversion. In fact, everything we use are these mappings from sites to weights.

File: src/projection_mpo.jl [hunk 13]
```diff
@@ -199,24 +207,213 @@ function projection_entries(::Type{T}, constraint::AbstractConstraint) where {T}
...
+
+  weights_by_site, rhs = sum_constraint_projection_data(constraint)
+  state_limit = sum_projection_state_limit(constraint.relation, rhs)
+  states_after_site = sum_projection_reachable_states(
+    weights_by_site,
+    length(sites),
+    constraint.relation,
+    rhs,
+  )
+  any(isempty, states_after_site) && return tensor_to_mpo(T, SparseTensorEntry{T}[], sites)
```
Comment (iagoleal) [inline src/projection_mpo.jl]:
```suggestion
  if any(isempty, states_after_site) 
    return tensor_to_mpo(T, SparseTensorEntry{T}[], sites)
  end
```

**Review by iagoleal — CHANGES_REQUESTED** [review]

**Comment by iagoleal** [comment]
An important design point that we should keep in mind: In the CoTenN paper, they build the projection MPOs by defining a finite automaton and then using a generic DFA to MPO builder. In here, we are defining a direct building method. This is ok, but also means that we will have to repeat a lot of code for each constraint.
To simplify further constraint building (including possible future new constraints outside the scope of #55), I think we should use their abstraction and define a `DFA` type and a generic MPO builder. From that, each constraint can be built declaratively by writing its DFA. This could even be a single method `constraint_dfa(constraint::SumConstraint)`, for example.

Having a generic construction can also be helpful when expanding the API or adding new backends, such as in #72 or #67.

**Review by bernalde — COMMENTED** [review]
Senior maintainer review of PR #86 at `d6d11ed2f94744a955bb49472700518d5bde0fcf`.

Verdict mechanics: I am the PR author as authenticated `bernalde`, so GitHub will not accept a formal APPROVE or REQUEST_CHANGES from this account. Submitting this as COMMENT. Substantive verdict: changes required.

Blocking: Documentation currently fails to build. 

The latest Documentation workflow fails on the reviewed head in `Build and deploy`: Documenter cannot resolve ``[`Integer`](@ref)`` from the `projection_mpo` docstring expanded into `docs/src/api.md`. The same doc block also needs a wording pass: `rhs` is not validated as nonnegative, and the implementation allocates states `0:rhs+1`, so the maximum state count is not simply `k`. Please replace the unresolved ref with plain text or a resolvable link and make the SumConstraint bond-dimension/RHS wording match the code.

Blocking: Non-integer `SumConstraint` projections silently use the generic path. 

`constraint_to_dfa(::SumConstraint{S}, sites)` only specializes `S <: Integer`, but there is no broader `SumConstraint` fallback. As a result, `SumConstraint([1], [0.5], :(<=), 1.0)` falls through to `constraint_to_dfa(::AbstractConstraint, sites)` and succeeds by enumerating feasible assignments. That contradicts the issue/PR contract that SumConstraint projection lowering is integer-only, and it removes the clear failure mode the earlier tests covered. Add an explicit `constraint_to_dfa(::SumConstraint, sites)` fallback that either validates integer-valued real inputs intentionally or throws `ArgumentError` for non-integer SumConstraint projections, with a regression test for fractional weights/RHS.

Blocking: The saturated integer DFA uses unchecked arithmetic at the bounds. 

The SumConstraint DFA computes `overflow = rhs + one(S)` and transitions with `q + weight * S(a)`. For accepted integer constraints near the type limit, this wraps before saturation. In a scratch load of `constraints.jl`/`projection_mpo.jl`, `is_feasible([1], SumConstraint([1], [typemax(Int)], :(>=), 1))` is true, but `constraint_to_dfa` throws `ArgumentError("transition table 1: unknown target state -9223372036854775808")`; with `rhs == typemax(Int)`, it throws `ArgumentError("states must not be empty")`. Please use checked arithmetic or wider arithmetic for the saturation calculation, or reject weights/RHS that cannot be represented safely with a clear `ArgumentError`, and restore boundary tests.

Nonblocking: none beyond the blocking items above.

Questions: none.

Tests and checks:
- Local diff file set against `origin/main` matched `gh pr diff --name-only`: `docs/src/api.md`, `src/constraints.jl`, `src/projection_mpo.jl`, `test/constraints.jl`, `test/projection_mpo.jl`.
- GitHub Run tests workflow is green for this head across Julia lts/1/pre on Ubuntu, Windows, and macOS.
- GitHub Documentation workflow is failing for this head with the unresolved Documenter cross-reference above.
- Local projection-only scratch test passed: `test/projection_mpo.jl` reported 174/174 passing when loading `constraints.jl` and `projection_mpo.jl` directly.
- Local `using TenSolver` did not run because the ignored local `Manifest.toml` is stale and resolves `QUBODrivers v0.3.4`; `main` has the same load-time `honors_final_reads` reference, while GitHub resolves `QUBODrivers v0.6.5`, so I treated that as environment noise.

Merge-readiness: not merge-ready. The docs check is red and the SumConstraint projection contract has two correctness gaps. I would not merge this until the blocking issues above are addressed.

**Comment by iagoleal** [comment]
> Blocking: The saturated integer DFA uses unchecked arithmetic at the bounds. 

Removing these checks was an explicit decision. They make the code slower and hard to read and do not add anything on real use cases. If we are using Int, we know variables might overflow. Also, even when it does overflows (which is easy to predict from the `rhs` itself), it would imply constructing an MPO with bond dimension of ~2^63. That is completely infeasible in any real machine.
I am open to discussing this, though, in case there are other advantages from the checks.

**Review by iagoleal — APPROVED** [review]

**Comment by iagoleal** [comment]
@bernalde, do want to pass it through one last review / sanity check?
If so, I can wait to merge.

**Comment by bernalde** [comment]
Commits pushed:
- `ddbba32` `fix: clean SumConstraint validation whitespace`

Main changes made:
- Cleaned the trailing whitespace left in `validate_sites`, so the PR diff passes `git diff --check`.
- Verified the current `6dec4c2`/`ddbba32` code addresses the body-only review findings for the docs cross-reference/wording and fractional SumConstraint validation: the unresolved `Integer` doc ref is gone, SumConstraint weights/RHS are validated as integer-valued, and fractional weight/RHS construction now throws `ArgumentError`.

Comments intentionally not addressed:
- The overflow-boundary guard from `<!-- gh-review-pr:finding=sumconstraint-overflow -->` remains intentionally not implemented. @iagoleal explicitly called this out as a deliberate tradeoff because such states would require infeasible MPO bond dimensions, then approved the PR at `6dec4c2`. I’m treating that as maintainer-accepted risk rather than pushing a guard that conflicts with the approved design.

Tests run:
- `git diff --check origin/main...HEAD` passed after `ddbba32`.
- Local targeted scratch test passed: `test/projection_mpo.jl` reported 238/238 passing when loading `constraints.jl` and `projection_mpo.jl` directly.
- Local fractional validation check passed: `SumConstraint([1], [0.5], :(<=), 1.0)` and `SumConstraint([1], [1.0], :(<=), 0.5)` both throw `ArgumentError`.
- GitHub checks for `ddbba32` passed after one rerun: Documentation passed, and the full Run tests matrix passed. The first test attempt had an unrelated `docs/src/examples.md` graph-coloring doctest representative flip outside this PR's diff; the rerun passed cleanly.

Remaining risks or approval gates:
- PR #86 retains formal maintainer approval and GitHub reports the branch merge state as `CLEAN`.

**Review by iagoleal — APPROVED** [review]
